tools/websoc-scraper-v2/index.ts

```diff
@@ -0,0 +1,453 @@
+ */
+type EnhancedSection = {
+  school: WebsocSchool;
+  department: WebsocDepartment;
+  course: WebsocCourse;
+  section: WebsocSection;
+};
+
+/**
+ * Object that contains all relevant data for a term.
+ */
+type ScrapedTerm = {
+  /**
+   * All ``WebsocAPIResponses`` returned by querying a department.
+   */
+  department: Record<string, WebsocAPIResponse>;
+  /**
+   * All ``WebsocAPIResponses`` returned by querying a GE category.
+   */
+  ge: Record<string, WebsocAPIResponse>;
+};
+
+/**
+ * An instructor object that can be inserted directly into Prisma.
+ */
+type ProcessedInstructor = {
+  year: string;
+  quarter: Quarter;
+  sectionCode: number;
+  timestamp: Date;
+  name: string;
+};
+
+/**
+ * A meeting object that can be inserted directly into Prisma.
+ */
+type ProcessedMeeting = {
+  year: string;
+  quarter: Quarter;
+  sectionCode: number;
+  timestamp: Date;
+  days: string[];
+  buildings: string[];
+  startTime: number;
+  endTime: number;
+};
+
+/**
+ * A section object containing additional metadata.
+ */
+type ProcessedSection = {
+  /**
+   * Contains the instructors and meetings associated with this section.
+   */
+  meta: {
+    instructors: ProcessedInstructor[];
+    meetings: ProcessedMeeting[];
+  };
+  /**
+   * The section object that can be inserted directly into Prisma.
+   */
+  data: {
+    year: string;
+    quarter: Quarter;
+    sectionCode: number;
+    timestamp: Date;
+    geCategories: GE[];
+    department: string;
+    courseNumber: string;
+    courseNumeric: number;
+    courseTitle: string;
+    sectionType: (typeof sectionTypes)[number];
+    units: string;
+    maxCapacity: number;
+    sectionFull: boolean;
+    waitlistFull: boolean;
+    overEnrolled: boolean;
+    cancelled: boolean;
+    data: object;
+  };
+};
+
+// The duration to sleep between scraping runs, or to wait before querying the
+// database for terms to scrape on demand if there were none in the last query.
+const SLEEP_DURATION = 5 * 60 * 1000; // 5 minutes
+
+const prisma = new PrismaClient();
+
+const logger = createLogger({
+  level: "info",
+  format: format.combine(
+    process.env.NODE_ENV === "development"
+      ? format.colorize({ all: true })
+      : format.uncolorize(),
+    format.timestamp(),
+    format.printf((info) => `${info.timestamp} [${info.level}] ${info.message}`)
+  ),
+  transports: [new transports.Console()],
+  exitOnError: false,
+});
+
+/**
+ * Sleep for the given number of milliseconds.
+ * @param duration Duration in ms.
+ */
+const sleep = async (duration: number) =>
+  new Promise((resolve) => setTimeout(resolve, duration));
+
+/**
+ * Get all terms that are to be scraped on a daily basis.
+ * @param date The current date.
+ */
+async function getTermsToScrape(date: Date) {
+  const quarterDates = Object.assign(
+    {},
+    ...(await Promise.all(
+      Array.from(Array(3).keys()).map((x) =>
+        getTermDateData((x + date.getFullYear() - 1).toString())
+      )
+    ))
+  );
```

Comment:
```suggestion
  const termDateData = await Promise.all(
    [
      ...(Array(3).keys()).map((x) =>
        getTermDateData((x + date.getFullYear() - 1).toString()
      )
    ])
  const quarterDates = {...termDateData}
```
I'm not sure of the reasoning behind the original code, but I was wondering if it could be simplified to something like this?